### ISSUE TITLE
Feature Request: pareto.tcl Design Space Exploration (Pareto Front) script

### ISSUE THREAD

**Issue by oharboe** [The-OpenROAD-Project/OpenROAD-flow-scripts#3738]
### Description

Dealing with death by knobs... Adding one more knob for any one feature seems like a small additional cost, but it is actually exponentially worse: it adds a dimension. One bad knob can cripple the usability of a feature.

## Framework inversion

Avoid framework inversion. Dont require a specific DSE frameworks by making sure the pareto.tcl script can run standalone and be used in any context of users choice: bazel, ray, optuna, home brew, etc.

the pareto.tcl below should also exist for each major stage as it makes sense. pareto-synth.tcl would be adequate to explore RTL parameterization and find a reasonable clock period/area range. This can help reduce dimensionality of DSE and reduce running times from hours, days weeks down to minutes and hours. Also, some stages are inheritely serial and others inheritely parallel, so provisioning becomes easier if synthesis(serial) can be handled in a stage separete from gpl(parallel).

## Motivation: The Hierarchy of "Knobs"

In the current state of physical design, users are often overwhelmed by configuration variables. We can classify design "knobs" into a hierarchy of usability:

1. **Tier 1 (Ideal): No Knob.** The tool is smart enough to find the optimal setting automatically.
2. **Tier 2 (Acceptable): Observable Knob.** The user sets a value, and the resulting report provides clear feedback (e.g., "congestion high," "timing violated"), allowing the user to infer the correct adjustment for the next run.
3. **Tier 3 (The Problem): Magic Knob.** The user must set a value (e.g., `CTS_CLUSTER_SIZE`, `ABC_AREA_SCORE`, `PLACE_DENSITY`) but has no intuition for the "right" number. The reports do not offer clues on how to adjust it. This leads to blind guessing.

ORFS currently possesses several Tier 3 knobs. The "Alternative" to the user guessing these numbers is to automate the guessing process entirely.

## Proposal

I propose a native mechanism for **Design Space Exploration (DSE)** to generate a Pareto front for **Minimum Clock Period**, **Minimum Area**, and **Minimum Power**.

This feature would allow a user to "run this overnight" and wake up to a set of optimized configurations rather than a single, manually guessed design point.

### 1. New Target: `make do-pareto`

A new Make target that acts as a wrapper for a hyperparameter optimization engine (conceptually similar to Optuna).

* **Input:** None. ORFS knows ranges and variables..
* **Execution:** Spawns parallel jobs to maximize hardware utilization.
* **Output:** A summary report containing the Pareto front metrics, Optuna provides a nice pdf pareto front report, which I presume any suitable framework will do.

### 2. Specialized Script: `pareto.tcl`

A modified flow script optimized specifically for DSE speed and low overhead.

* **Lightweight Output:** Unlike the standard flow, this should **not** write heavy artifacts (`.odb`, `.def`, `.gds`) to disk for every trial. It should only emit the final metrics (logs/reports).
* **Fast Fail:** Implements aggressive "pruning." If `global_placement` yields high overflow or `synthesis` yields unfixable WNS, the script aborts immediately to save compute time for the next trial.
* **Metric Extraction:** Automatically calculates the "Effective Clock Period" (Target Period - WNS) to normalize results across different frequency targets.
* **accuracy redefined** The actual PPA numbers are unobservable(not an output) of this flow, the output is the pareto front of the knobs, the values of the knobs. So only directional accuracy of PPA matters for the pareto.tcl script. Use this to speed up the script and increase parallelism, if possible.

### 3. Integration with Optuna(or better)

The backend should leverage a standard optimizer like Optuna to intelligently sample the design space, rather than a simple grid search.

* **Objectives:** Multi-objective optimization (minimize Area, minimize Power, minimize Effective Period).
* **Storage:** Support for SQL/storage backends to allow resuming studies or distributed execution.

## User Workflow Example

Instead of manually tweaking `config.mk` 10 times:

```bash
# User defines ranges in a simple config or accepts defaults
# User runs the exploration
make do-pareto N_TRIALS=100 JOBS=8

# ORFS runs 100 designs intelligently.
# Returns: "Best Speed config: ...", "Best Area config: ..."

```

## Benefits

* **Democratization:** Allows non-experts to achieve expert-level PPA by trading compute time for manual tuning time.
* **Efficiency:** Prevents disk bloat by discarding artifacts from sub-optimal intermediate trials.
* **Data-Driven:** Moves ORFS from a "static flow" to an "optimizing engine."

## Managing expectations

This will work for simple designs like the ones in the ORFS test database and is great for training/onboarding and generating test cases in ORFS/OpenROAD

For more advanced DSE, such as that involving RTL parameterization(maximum compute/energy in a system where there is out of band static power draw), the user will have to implement his own optimization function, flow and  Python script to run the DSE for full flexibility.

**Comment by oharboe** [comment]
@maliberty @MrAMS This is more of a write-up/planning than a feature request as such. It describes what doesn't exist today but we'd like to see for ORFS, given infinite resources, so that we're can plan better how to work around this and what not to pursue as it would be engulfed by this broader vision.

**Comment by oharboe** [comment]
@tspyrou FYI

**Comment by maliberty** [comment]
This is a description of the auto-tuner.  I know you don't like it so I don't know what you are requesting here that is different.

**Comment by oharboe** [comment]
Usability, fix framework inversion. Listed above.

Also, this is a nice feature in theory, but as I describe, it doesnt cover my DSE use case.

I am glad you think it is a reasonable description of AutoTuner and some known issues that are not fixed for lack of cycles. I was asked what AutoTuner is compared to DSE and I wrote this up in response.

**Comment by maliberty** [comment]
How is this script going to know how to run all these different flow?  It seems it will take a lot of configuration for it to be usable.  For any non-trivial number of knobs and design size it will need to be distributed and TCL is a poor choice for that.

**Comment by oharboe** [comment]
@maliberty Please close this issue if it contains nothing new or helpful.

**Comment by oharboe** [comment]
I am looking for .tcl tidbits that are useful in my flow, not a framework or dependency distribution choice made for me

**Comment by oharboe** [comment]
If this is undoable, then my feature request has been satisfactorily answered: I have to roll my own optimization function, which I would like to have useful tidbits for, but nothing beyond what is ORFS scripts today is practically meaninful. I have to roll my own.

**Comment by oharboe** [comment]
Here is a PR merged that is a small nice step towards more composable flow scripts https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/3734

**Comment by maliberty** [comment]
I'm open to helping but I don't know what sort of tidbit you are requesting.  What will be the advantage of your own version of AT?  Will it be the inversion you describe above?  If not, how will it differ?
